src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php

```diff
@@ -52,6 +52,12 @@ protected function processValue($value, $isRoot = false)
             if (!$v instanceof Reference) {
                 throw new InvalidArgumentException(sprintf('Invalid definition for service "%s": an array of references is expected as first argument when the "container.service_locator" tag is set, "%s" found for key "%s".', $this->currentId, \is_object($v) ? \get_class($v) : \gettype($v), $k));
             }
+
+            if (\is_int($k)) {
+                unset($arguments[0][$k]);
+
+                $k = (string) $v;
```

Comment:
should we check whether this key already exists in the configured arguments ?

Reply:
I'm not sure: if the key exists *after* the numbered one, it will override it - and if it came before, then the numbered one should win to me. works for you also?

Reply:
nope, it will *always* override it here. If it comes after the numbered one, processing it will see the value set by the processing of the numbered one, not the original one, as it would have been replaced

Reply:
Really? foreach operates on a copy of array so that it will yield the original value at this iteration, and overwrite the value set by the numbered step to me.

Reply:
Added some basic tests that should confirm @nicolas-grekas description. I do like the idea of preventing unexepected behavior for the user but it works either way for me.